internal/agent/runtime/docker.go

```diff
@@ -0,0 +1,191 @@
+	// on the action are therefore the command portion in Docker.
+	if a.Cmd != "" {
+		cfg.Entrypoint = append(cfg.Entrypoint, a.Cmd)
+	}
+	if len(a.Args) > 0 {
+		cfg.Cmd = append(cfg.Cmd, a.Args...)
+	}
+
+	// TODO: Figure out container logging. We probably want to save it somewhere for debugability.
+
+	create, err := d.client.ContainerCreate(ctx, &cfg, &hostCfg, nil, nil, containerName)
+	if err != nil {
+		return fmt.Errorf("docker: %w", err)
+	}
+
+	// Always try to remove the container on exit.
+	defer func() {
+		// We can't use the context passed to Run() as it may have been cancelled.
+		err := d.client.ContainerRemove(context.Background(), create.ID, types.ContainerRemoveOptions{})
+		if err != nil {
+			d.log.Info("Couldn't remove container", "container_name", containerName, "error", err)
+		}
+	}()
+
+	// Issue the wait with a 'next-exit' condition so we can await a response originating from
+	// ContainerStart().
+	waitBody, waitErr := d.client.ContainerWait(ctx, create.ID, container.WaitConditionNextExit)
+
+	if err := d.client.ContainerStart(ctx, create.ID, types.ContainerStartOptions{}); err != nil {
+		return fmt.Errorf("docker: %w", err)
+	}
+
+	select {
+	case result := <-waitBody:
+		if result.StatusCode == 0 {
+			return nil
+		}
+		return failureFiles.ToError()
+
+	case err := <-waitErr:
+		return fmt.Errorf("docker: %w", err)
+
+	case <-ctx.Done():
+		// We can't use the context passed to Run() as its been cancelled.
+		err := d.client.ContainerStop(context.Background(), create.ID, container.StopOptions{
+			Timeout: ptr.Int(5),
+		})
+		if err != nil {
+			d.log.Info("Failed to gracefully stop container", "error", err)
+		}
+		return fmt.Errorf("docker: %w", ctx.Err())
+	}
+}
+
+var validContainerName = regexp.MustCompile(`[^a-zA-Z0-9_.-]`)
+
+// toContainerName converts an action ID into a usable container name.
+func toContainerName(actionID string) string {
+	// Prepend 'tinkerbell_' so we guarantee the additional constraints on the first character.
+	return "tinkerbell_" + validContainerName.ReplaceAllString(actionID, "_")
+}
+
+func toDockerEnv(env map[string]string) []string {
+	var de []string
+	for k, v := range env {
+		de = append(de, fmt.Sprintf("%v=%v", k, v))
+	}
+	return de
+}
+
+// NewDocker creates a new Docker instance.
+func NewDocker(opts ...DockerOption) (*Docker, error) {
+	o := &Docker{}
+
+	for _, fn := range opts {
+		fn(o)
+	}
+
+	if o.log.GetSink() == nil {
+		o.log = logr.Discard()
+	}
+
+	if o.client == nil {
+		clnt, err := client.NewClientWithOpts(client.FromEnv)
+		if err != nil {
+			return nil, err
+		}
+		clnt.NegotiateAPIVersion(context.Background())
+		o.client = clnt
+	}
+
+	return o, nil
+}
+
+// DockerOption defines optional configuration for a Docker instance.
+type DockerOption func(*Docker)
+
+// WithLogger returns an option to configure the logger on a Docker instance.
+func WithLogger(log logr.Logger) DockerOption {
```

Comment:
I've realized the logger is actually required because we log when a container couldn't be gracefully stopped. I'll tweak this when adding the other features as noted in the OP.